include/picongpu/particles/creation/SpawnFromSourceSpecies.kernel

```diff
@@ -0,0 +1,402 @@
+                         &productSpeciesFrameArray](uint32_t const linearIdx)
+                        {
+                            int32_t const numberFreeSlotsPreviousLinearIdx
+                                = static_cast<int32_t>(linearIdx * frameSize) - offsetLowFrame;
+                            // for low will always be <0(partially filled) or ==0(nullptr)
+
+                            if(
+                                // need slots from frame, since previous is not enough
+                                (numberFreeSlotsPreviousLinearIdx
+                                 < static_cast<int32_t>(totalNumberProductSpeciesParticlesToSpawn))
+                                &&
+                                // is nullPtr
+                                !(productSpeciesFrameArray[linearIdx].isValid()))
+                            {
+                                productSpeciesFrameArray[linearIdx] = productBox.getEmptyFrame(worker);
+                                productBox.setAsLastFrame(worker, productSpeciesFrameArray[linearIdx], superCellIndex);
+                            }
+                        });
+                    worker.sync();
+
+                    // try to init one productParticle for each logical worker
+                    forEachFrameSlot(
+                        [&worker,
+                         &sourceSpeciesFrame,
+                         &productSpeciesFrameArray,
+                         &frameSize,
+                         &spawnCounter,
+                         &offsetLowFrame,
+                         &sharedDataBoxes...](uint32_t const frameSlotIdx, TypeNumber& numberProductParticles)
+                        {
+                            auto sourceParticle = sourceSpeciesFrame[frameSlotIdx];
+
+                            // not occupied or does not want to spawn a particle
+                            if(numberProductParticles == static_cast<TypeNumber>(0u))
+                                return;
+
+                            uint32_t const globalSlotIndex = alpaka::atomicAdd(
+                                worker.getAcc(),
+                                &spawnCounter,
+                                static_cast<uint32_t>(1u),
+                                ::alpaka::hierarchy::Threads{});
+
+                            uint32_t const productSpeciesFrameIndex = (globalSlotIndex + offsetLowFrame) / frameSize;
+                            uint32_t const localSlotIndex = (globalSlotIndex + offsetLowFrame) % frameSize;
+
+                            auto productParticle = productSpeciesFrameArray[productSpeciesFrameIndex][localSlotIndex];
+
+                            // init frame slot, mark as valid particle
+                            productParticle[multiMask_] = 1u;
+
+                            // init productParticle and maybe change source particle attributes
+                            if constexpr(T_passCascadeIndex && T_passSharedDataBoxes)
```

Comment:
please call always the module functors. A functor can be empty if needed.

Reply:
Done